File: Legacy/TextFormatter.vb
Imports System.Collections.Generic

Public Class TextFormatter

    ' the Player generates style tags for us
    ' so all we need to do is have some kind of <color> <fontsize> <justify> tags etc.
    ' it would actually be a really good idea for the player to handle the <wait> and <clear> tags too...?

    Private bold As Boolean
    Private italic As Boolean
    Private underline As Boolean
    Private colour As String = ""
    Private fontSize As Integer = 0
    Private align As String = ""

    Public Function OutputHTML(ByVal input As String) As String
        Dim output As String = ""
        Dim position As Integer = 0
        Dim codePosition As Integer
        Dim finished As Boolean = False

        input = input.Replace("<", "&lt;").Replace(">", "&gt;").Replace(vbCrLf, "<br />")

        Do
            codePosition = input.IndexOf("|", position)
            If codePosition = -1 Then
                output += FormatText(input.Substring(position))
                finished = True
            Else
                output += FormatText(input.Substring(position, codePosition - position))
                position = codePosition + 1

                Dim oneCharCode As String = ""
                Dim twoCharCode As String = ""
                If position < input.Length Then
                    oneCharCode = input.Substring(position, 1)
                End If
                If position < (input.Length - 1) Then
                    twoCharCode = input.Substring(position, 2)
                End If

                Dim foundCode As Boolean = True

                Select Case twoCharCode
                    Case "xb"
                        bold = False
                    Case "xi"
                        italic = False
                    Case "xu"
                        underline = False
                    Case "cb"
                        colour = ""
                    Case "cr"
                        colour = "red"
                    Case "cl"
                        colour = "blue"
                    Case "cy"
                        colour = "yellow"
                    Case "cg"
                        colour = "green"
                    Case "jl"
                        align = ""
                    Case "jc"
                        align = "center"
                    Case "jr"
                        align = "right"
                    Case Else
                        foundCode = False
                End Select

                If foundCode Then
                    position += 2
                Else
                    foundCode = True
                    Select Case oneCharCode
                        Case "b"
                            bold = True
                        Case "i"
                            italic = True
                        Case "u"
                            underline = True
                        Case "n"
                            output += "<br />"
                        Case Else
                            foundCode = False
                    End Select

                    If foundCode Then
                        position += 1
                    End If
                End If

                If Not foundCode Then
                    If oneCharCode = "s" Then
                        ' |s00 |s10 etc.
                        If position < (input.Length - 2) Then
                            Dim sizeCode As String = input.Substring(position + 1, 2)
                            If Integer.TryParse(sizeCode, fontSize) Then
                                foundCode = True
                                position += 3
                            End If
                        End If
                    End If
                End If

                If Not foundCode Then
                    output += "|"
                End If

                ' can also have size codes

            End If

        Loop Until finished Or position >= input.Length

        Return output

    End Function

    Private Function FormatText(ByVal input As String) As String
        If input.Length = 0 Then Return input

        Dim output As String = ""

        If align.Length > 0 Then output += "<align align=""" + align + """>"
        If fontSize > 0 Then output += "<font size=""" + CStr(fontSize) + """>"
        If colour.Length > 0 Then output += "<color color=""" + colour + """>"
        If bold Then output += "<b>"
        If italic Then output += "<i>"
        If underline Then output += "<u>"
        output += input
        If underline Then output += "</u>"
        If italic Then output += "</i>"
        If bold Then output += "</b>"
        If colour.Length > 0 Then output += "</color>"
        If fontSize > 0 Then output += "</font>"
        If align.Length > 0 Then output += "</align>"

        Return output
    End Function

End Class

File: ASL4/TextFormatter.cs
using Microsoft.VisualBasic;

namespace TextAdventures.Quest.LegacyASL
{
    public class TextFormatter
    {
        // the Player generates style tags for us
        // so all we need to do is have some kind of <color> <fontsize> <justify> tags etc.
        // it would actually be a really good idea for the player to handle the <wait> and <clear> tags too...?

        private bool bold;
        private bool italic;
        private bool underline;
        private string colour = "";
        private int fontSize = 0;
        private string align = "";

        public string OutputHTML(string input)
        {
            string output = "";
            int position = 0;
            int codePosition;
            bool finished = false;
            bool nobr = false;

            input = input.Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;").Replace(Constants.vbCrLf, "<br />");

            if (Strings.Right(input, 3) == "|xn")
            {
                nobr = true;
                input = Strings.Left(input, Strings.Len(input) - 3);
            }

            do
            {
                codePosition = input.IndexOf("|", position);
                if (codePosition == -1)
                {
                    output += FormatText(input.Substring(position));
                    finished = true;
                }
                else
                {
                    output += FormatText(input.Substring(position, codePosition - position));
                    position = codePosition + 1;

                    string oneCharCode = "";
                    string twoCharCode = "";
                    if (position < input.Length)
                    {
                        oneCharCode = input.Substring(position, 1);
                    }
                    if (position < input.Length - 1)
                    {
                        twoCharCode = input.Substring(position, 2);
                    }

                    bool foundCode = true;

                    switch (twoCharCode ?? "")
                    {
                        case "xb":
                            {
                                bold = false;
                                break;
                            }
                        case "xi":
                            {
                                italic = false;
                                break;
                            }
                        case "xu":
                            {
                                underline = false;
                                break;
                            }
                        case "cb":
                            {
                                colour = "";
                                break;
                            }
                        case "cr":
                            {
                                colour = "red";
                                break;
                            }
                        case "cl":
                            {
                                colour = "blue";
                                break;
                            }
                        case "cy":
                            {
                                colour = "yellow";
                                break;
                            }
                        case "cg":
                            {
                                colour = "green";
                                break;
                            }
                        case "jl":
                            {
                                align = "";
                                break;
                            }
                        case "jc":
                            {
                                align = "center";
                                break;
                            }
                        case "jr":
                            {
                                align = "right";
                                break;
                            }

                        default:
                            {
                                foundCode = false;
                                break;
                            }
                    }

                    if (foundCode)
                    {
                        position += 2;
                    }
                    else
                    {
                        foundCode = true;
                        switch (oneCharCode ?? "")
                        {
                            case "b":
                                {
                                    bold = true;
                                    break;
                                }
                            case "i":
                                {
                                    italic = true;
                                    break;
                                }
                            case "u":
                                {
                                    underline = true;
                                    break;
                                }
                            case "n":
                                {
                                    output += "<br />";
                                    break;
                                }

                            default:
                                {
                                    foundCode = false;
                                    break;
                                }
                        }

                        if (foundCode)
                        {
                            position += 1;
                        }
                    }

                    if (!foundCode)
                    {
                        if (oneCharCode == "s")
                        {
                            // |s00 |s10 etc.
                            if (position < input.Length - 2)
                            {
                                string sizeCode = input.Substring(position + 1, 2);
                                if (int.TryParse(sizeCode, out fontSize))
                                {
                                    foundCode = true;
                                    position += 3;
                                }
                            }
                        }
                    }

                    if (!foundCode)
                    {
                        output += "|";
                    }

                    // can also have size codes

                }
            }

            while (!(finished | position >= input.Length));

            return string.Format("<output{0}>{1}</output>", nobr ? " nobr=\"true\"" : "", output);

        }

        private string FormatText(string input)
        {
            if (input.Length == 0)
                return input;

            string output = "";

            if (align.Length > 0)
                output += "<align align=\"" + align + "\">";
            if (fontSize > 0)
                output += "<font size=\"" + fontSize.ToString() + "\">";
            if (colour.Length > 0)
                output += "<color color=\"" + colour + "\">";
            if (bold)
                output += "<b>";
            if (italic)
                output += "<i>";
            if (underline)
                output += "<u>";
            output += input;
            if (underline)
                output += "</u>";
            if (italic)
                output += "</i>";
            if (bold)
                output += "</b>";
            if (colour.Length > 0)
                output += "</color>";
            if (fontSize > 0)
                output += "</font>";
            if (align.Length > 0)
                output += "</align>";

            return output;
        }

    }
}